### PR TITLE
Add DEC/IDEC tensorflow implemenetation

### PR DESCRIPTION
**Pull request by nickmarton**
This commit adds our re-implementation of DEC/IDEC on the MNIST dataset.

### CODE REVIEW

File: tf-IDEC/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+---
+The modules in this directory implement:
+
+* [DEC](http://proceedings.mlr.press/v48/xieb16.pdf) (Unsupervised Deep Embedding for Clustering Analysis) and
+* [IDEC](https://xifengguo.github.io/papers/IJCAI17-IDEC.pdf) (Xifeng Guo, Long Gao, Xinwang Liu, Jianping Yin. Improved Deep Embedded Clustering with Local Structure Preservation. IJCAI 2017)
+
+Everything is written in pure `tensorflow`/`python` on only the MNIST dataset. Special thanks to the authors for providing a Keras implementation [here](https://github.com/XifengGuo/IDEC), which we base a lot of this on. Testing error in our implementation converges between 87-88%.
+
+The key differences between our implementation and the paper are as follows:
+* We train our autoencoder end-to-end; that is, we do not to greedy, layer-wise training.
```
Comment (kborer) [inline tf-IDEC/README.md]: do

File: tf-IDEC/constants.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+"""Provide constants used throughout DEC/IDEC."""
+import os
+
+# Stopping tolerance
+tol= 1e-4
```
Comment (kborer) [inline tf-IDEC/constants.py]: spacing

File: tf-IDEC/idec.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        Compute soft targets q_ij via Sudent's t-distribution.
+
+        Here we compute the numerator of equation (1) for q_ij, then normalize
+        by dividing by the total sum over all vectors in the numerator.
+        """
+
+        # We use axis arg to norm so that the tensor is treated as a batch of vectors.
+        num = (1.0 + tf.norm((tf.expand_dims(inputs, axis=1) - self.clusters), axis=2) / self.alpha)
+        num **= -((self.alpha + 1.0) / 2.0)
+        return num /  tf.reduce_sum(num)
```
Comment (kborer) [inline tf-IDEC/idec.py]: spacing

File: tf-IDEC/idec.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        cross_entropy = -tf.reduce_sum(placeholders["target"] * tf.log(cluster))
+        entropy = -tf.reduce_sum(placeholders["target"] * tf.log(placeholders["target"] + 0.00001))
+        Lc = kl_divergence = cross_entropy - entropy
+
+        Lr = tf.losses.mean_squared_error(inp, decode_fn)
+
+        loss = Lr + gamma * Lc
+
+
+    # Add optimzation ops to the graph, then reinitialize all trainable
```
Comment (kborer) [inline tf-IDEC/idec.py]: optimization

File: tf-IDEC/idec.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            delta_label = np.sum(y_pred != y_pred_last).astype(np.float32) / y_pred.shape[0]
+            y_pred_last = y_pred
+
+            if itr > 0 and delta_label < tol:
+                logging.info(
+                    "Reached tolerance threshold (delta: {} < tol: {}). "
+                    "Stopping training.".format(delta_label, tol))
+                break
+
+        # train on one batch at a time batch
```
Comment (kborer) [inline tf-IDEC/idec.py]: one_batch_two_batch

**Review by kborer — COMMENTED** [review]
looks good. I haven't finished the papers yet, so see if @jiayukeyme can check the implementation details

File: tf-IDEC/README.md (hunk 1 above)
Comment (nickmarton) [reply tf-IDEC/README.md]: fixed

File: tf-IDEC/constants.py (hunk 2 above)
Comment (nickmarton) [reply tf-IDEC/constants.py]: fixed

File: tf-IDEC/idec.py (hunk 3 above)
Comment (nickmarton) [reply tf-IDEC/idec.py]: fixed

File: tf-IDEC/idec.py (hunk 4 above)
Comment (nickmarton) [reply tf-IDEC/idec.py]: fixed

File: tf-IDEC/idec.py (hunk 5 above)
Comment (nickmarton) [reply tf-IDEC/idec.py]: fixed

**Comment by nickmarton** [comment]
Forgot to add 
```
if __name__ == "__main__":
    main()
```
to `autoencoder.py`; added

File: tf-IDEC/constants.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+"""Provide constants used throughout DEC/IDEC."""
+import os
+
+# Stopping tolerance
+tol = 1e-3
```
Comment (jiayukeyme) [inline tf-IDEC/constants.py:5]: Maybe these constants are also uppercase?

File: tf-IDEC/autoencoder.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        np.random.shuffle(X)
+
+        # Train over all full batches; we shuffle so we don't have to worry
+        # about when a partial batch is left over
+        train_loss = 0
+        for batch_ix in range(num_batches):
+            # Get next batch for training, then reshape to be 4D
+            # (our first layer after the input layer will flatten the input)
+            batch_X = X[batch_ix * Nm: (batch_ix + 1) * Nm]
+            batch_X = np.reshape(batch_X, (Nm, NUM_CHANNELS, *IMAGE_SHAPE))
```
Comment (jiayukeyme) [inline tf-IDEC/autoencoder.py]: I might have missed something but X is already reshaped in `load_mnist()`

**Review by jiayukeyme — COMMENTED** [review]
Great work! Some minor issues, otherwise lgtm

File: tf-IDEC/constants.py (hunk 6 above)
Comment (nickmarton) [reply tf-IDEC/constants.py:5]: paper uses italicized, lowercase script for "*tol*", which is why it's lowercase

File: tf-IDEC/autoencoder.py (hunk 7 above)
Comment (nickmarton) [reply tf-IDEC/autoencoder.py]: Good call, removed and updated comment

File: tf-IDEC/constants.py (hunk 6 above)
Comment (jiayukeyme) [reply tf-IDEC/constants.py:5]: got it. lgtm

**Pull request merged**